### PR TITLE
webView: Consider message list height to detect whether list is short or not and get last message node visible on screen.

### PR DESCRIPTION
**Pull request by jainkuniya**

### CODE REVIEW

**Comment by borisyankov** [comment]
I used this `wrapper` approach a few months ago, to solve a different problem.
There were several issues that this caused, some we might already be addressing or be more robust about them (was about messages, parents etc.).

One issue that I am not sure we can solve (maybe we can) is that the sticky header breaks, or more correctly sticks somewhere outside of view ;)

File: src/webview/js/generatedEs3.js [hunk 1]
```diff
@@ -142,7 +145,8 @@ var sendScrollMessage = function sendScrollMessage() {
 };
 
 var sendScrollMessageIfListShort = function sendScrollMessageIfListShort() {
-  if (documentBody.scrollHeight === documentBody.clientHeight) {
+  var messageListHeight = document.getElementById('message-list').clientHeight;
+  if (messageListHeight <= documentBody.clientHeight) {
```
Comment (gnprice) [inline src/webview/js/generatedEs3.js]: This doesn't seem right -- take a look at the documentation of `clientHeight` and `scrollHeight`, on MDN. We should be doing a comparison on the total height of the message list, aka its `scrollHeight`.

In particular I believe this condition will always be true. With a strict inequality (`<`), it would be true when any of the other children in the body, like the loading spinners, are on screen.

File: src/webview/js/generatedEs3.js [hunk 2]
```diff
@@ -116,11 +116,14 @@ var isMessageNode = function isMessageNode(node) {
 
 var getStartAndEndNodes = function getStartAndEndNodes() {
   var startNode = getMessageNode(document.elementFromPoint(200, 20));
-  var endNode = getMessageNode(document.elementFromPoint(200, window.innerHeight - 20));
+  var windowInnerHeight = window.innerHeight;
+  var messageListHeight = document.getElementById('message-list').clientHeight;
+  var endPoint = messageListHeight > 0 ? Math.min(windowInnerHeight, messageListHeight) : windowInnerHeight;
+  var endNode = getMessageNode(document.elementFromPoint(200, endPoint - 20));
```
Comment (gnprice) [inline src/webview/js/generatedEs3.js]: This logic isn't right -- in particular, you're counting by a distance `messageListHeight` down from the top of the whole viewport, but the message list may not start at the top of the viewport.

Take a look at [`getBoundingClientRect`](https://developer.mozilla.org/en-US/docs/Web/API/Element/getBoundingClientRect) -- I think it may save you from trying to sort through this by hand, and give shorter and clearer code.

File: src/webview/js/generatedEs3.js [hunk 3]
```diff
@@ -116,11 +116,14 @@ var isMessageNode = function isMessageNode(node) {
...
 
   return {
-    start: isMessageNode(startNode) ? startNode.getAttribute('data-msg-id') : Number.MAX_SAFE_INTEGER,
-    end: isMessageNode(endNode) ? endNode.getAttribute('data-msg-id') : 0
+    start: isMessageNode(startNode) ? +startNode.getAttribute('data-msg-id') : 1,
```
Comment (gnprice) [inline src/webview/js/generatedEs3.js]: Why does the fallback value change from `Number.MAX_SAFE_INTEGER` to 1? Those are pretty different values :wink:

**Comment by borisyankov** [comment]
@jainkuniya I was working on this and have a simpler fix for the issue, but paused to see what happens with this PR. Let's make sure we figure out if this approach will work, so if it doesn't I'll finish up mine :)

In particular, the sticky headers is a deal-breaker if you can't make it work with a wrapper.

**Comment by jainkuniya** [comment]
@borisyankov i got a way to get same benefit as this PR, without wrapper. Will try and update it tonight.

**Comment by gnprice** [comment]
Thanks @jainkuniya ! A few comments:
* This PR is kind of hard to follow, because the first commits go in one direction and then subsequent commits undo them. Please use `git rebase -i` to produce a branch that reflects what you think should be merged, with each commit individually coherent and each commit moving in the desired direction.
* I don't like the approach in `getMessageListHeight` where we try to think of all the different elements that will make up the message list, and add up their height. That's going to be super brittle as we make changes to the exact HTML in the message list.
* I'd be glad to see a commit that adds a div as the common ancestor for the whole message list. (A "wrapper", as it's been called in this thread.) That makes a clean solution to this problem and others, and in general is something I almost always see when looking at a web app's DOM. I think focusing on that problem, and solving whatever goes wrong with a first attempt to do it, would be quite productive.

**Comment by borisyankov** [comment]
@jainkuniya the PR I told you I was working on and paused... used a much simpler approach :)
I am sending the last message ID with the `content` message.

**Comment by gnprice** [comment]
Closing in favor of #2988 and the PRs referring to it.
